vdr/didstore/event.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/nuts-foundation/nuts-node/network/dag"
+)
+
+// event contains the transaction reference and ordering of all DID document diff
+type event struct {
+	// Created is the transaction creation time, used for sorting
+	Created time.Time `json:"created"`
+	// LogicalClock contains the LC header from the transaction
+	LogicalClock uint32 `json:"lc"`
+	// TXRef contains the TX.Ref of the original transaction. Used for ordering
+	TXRef hash.SHA256Hash `json:"txref"`
+	// DocRef contains the reference to the document shelf. Equals transaction payload hash
+	DocRef hash.SHA256Hash `json:"docref"`
+	// MetaRef contains a reference to the documentMetadata shelf. "DID + version"
+	MetaRef string `json:"metaref"`
+	// document contains the created Document. This needs to be added to a new event since we cannot write and read within the same TX.
+	document *did.Document
+	// metadata contains the created Metadata. This needs to be added to a new event since we cannot write and read within the same TX.
+	metadata *documentMetadata
+}
+
+// Len returns the length of the Events slice. Required for sorting.
+func (el *eventList) Len() int {
+	return len(el.Events)
+}
+
+// Less is part of the methods required for sorting
+func (el *eventList) Less(i, j int) bool {
+	left := el.Events[i]
+	right := el.Events[j]
+
+	if left.LogicalClock < right.LogicalClock {
+		return true
+	}
+	if left.LogicalClock > right.LogicalClock {
+		return false
+	}
+
+	return left.Created.Before(right.Created)
+}
+
+// Swap is part of the methods required for sorting
+func (el *eventList) Swap(i, j int) {
+	el.Events[i], el.Events[j] = el.Events[j], el.Events[i]
+}
+
+func eventFromTransaction(transaction dag.Transaction) event {
+	return event{
+		Created:      transaction.SigningTime(),
+		LogicalClock: transaction.Clock(),
+		TXRef:        transaction.Ref(),
+		DocRef:       transaction.PayloadHash(),
+	}
+}
+
+// eventList is an in-memory representation of an Events shelf entry
+type eventList struct {
+	Events []event `json:"events"`
+}
+
+func (el *eventList) copy() eventList {
+	cpy := eventList{Events: make([]event, len(el.Events))}
+	copy(cpy.Events, el.Events)
+
+	return cpy
+}
+
+// insert the event at the correct location
+func (el *eventList) insert(e event) {
+	// 1% case
+	if len(el.Events) == 0 {
+		el.Events = append(el.Events, e)
+		return
+	}
+
+	// 98.99% case
+	last := el.Events[len(el.Events)-1]
+	el.Events = append(el.Events, e)
+	if last.LogicalClock < e.LogicalClock {
+		return
+	}
+
+	// 0.01% case
+	sort.Stable(el)
+}
+
+// diff returns the latest matching event and a sublist of events that need to be applied to the latest.
+// Given transaction orderings A: 1->2->4 and B: 1->2->3.
+// A.diff(B) results in 2, [3,4] where 2 is the last common TX (base version of DID document)
+// and 3,4 are updates that have to be applied to the base version as updates.
+func (el *eventList) diff(updated eventList) (*event, []event) {
```

Comment:
`store.Add` only adds 1 `did.Document` at the time, so it looks a bit complicated. When we don't add multiple documents at once, you don't need `diff` and can always use `insert`?

Reply:
Updates are added one by one through transactions. They may come out of order. Add doesn't care about ordering, it'll fix it for you.